inst/training_202307/202307.qmd

```diff
@@ -0,0 +1,448 @@
+---
+title: "`verdepcheck`"
+author: "Pawel Rucki"
+date: 2023/08/07
+format: revealjs
+---
+
+## Agenda
+
+1. Motivation
+1. Solution
+1. How to use?
+1. How it works?
+1. Discussion
+1. Future work
+
+# Motivation
+
+```{.yaml code-line-numbers="7"}
+Type: Package
+Package: foo.package
+Version: 1.2.3.9000
+Depends:
+    R (>= 3.6)
+Imports:
+    dplyr,
+    (...)
+```
+
+The above specification indicates compatibility with **all** versions of `dplyr` which is very unlikely to be true.
+
+# Motivation
+
+```.default
+Error: object ‘foo’ is not exported by 'namespace:bar'
+```
+
+<br>
+
+```.default
+`foo()` was deprecated in bar 1.0.0.
+i Please use `baz()` instead.
+```
+
+<br>
+
+The new versions of dependencies might introduce a breaking changes.
+
+# Motivation
+
+R environments for clinical trials are usually not updated very frequently.
+
+Users might use `renv` for environment management and packages used might not get updates for some time.
+
+As a result, your package might be used with _older_ version of dependent packages.
+
+# Solution
+
+`verdepcheck` (**ver**sion of **dep**endencies **check**) - a tool for package developers to check your package against different set of dependencies. It is meant to be used primarily in CI.
+
+Available at:
+
+- [`insightsengineering/verdepcheck`](https://github.com/insightsengineering/verdepcheck)
+- [`insightsengineering/r-verdepcheck-action`](https://github.com/insightsengineering/r-verdepcheck-action)
+
+:::{.callout-important}
+Do not confuse `verdepcheck` with `revdepcheck` for reverse dependency checks. Those are totally different tools!
+:::
+
+#
+
+Let's call a "different set of dependencies" a _strategy_. Currently there are three _strategies_ implemented:
+
+- **max** - use _development_ version of dependencies
+- **release** - use _released_ version of dependencies (e.g. from CRAN)
+- **min** - use _minimal_ version of dependencies as per `DESCRIPTION` file
+
+This tool is looking into a **direct** dependencies only.
+
+## Explanation for direct dependencies
+
+The initial prototype of `min` strategy included recursive find of minimal dependency. This oftentimes lead into usage of already archived package or not able to compile very old package on modern systems architecture.
```

Comment:
This is a function of the system you're running your tests on, though, and not an inherent "code failure" in the sense that min is searching for. 

That said we should be using versions of recursive dependencies that were concurrent with the versions of the direct deps being tested. switchr has code to do this. You'll need to transform a switcher manifest to a pak manifest, but that should be easily doable. 

Testing old versions of direct dependencies with modern versions of their recursive dependencies is not a useful or realistic check, in my opinion.  Easy to get breakages just from that mismatch that dont' reflect any real cohort that anyone is actually going to have in their renv/locked site library

Reply:
> This is a function of the system you're running your tests on, though, and not an inherent "code failure" in the sense that min is searching for.

True. I tried to include that in the discussions down below about min deps putthing this as it's _installable_ set of dependencies - i.e. not necessarily code-wise minimal set of dependencies. I will try to rephrase that section again.

> That said we should be using versions of recursive dependencies that were concurrent with the versions of the direct deps being tested. switchr has code to do this.

That's a very interesting idea - thanks for putting this. I haven't thought about that this way. I will have to explore it. Maybe we can make a new strategy out of it? Just thinking... I am going to put this in the future work not to miss that.
The problem that I am seeing right now is that we would have to resolve (recursively) the whole dependency tree on our end. It's of course doable but from the very beginning I wanted to keep it as a somewhat thin wrapper on `pkgdepends` and relay on it when it comes to deps resolve, download, install and all of that. Putting custom steps into its workflow it's not an easy task and require so ugly workarounds (something that I did once here but I am not very proud of it). Having this fully custom also comes with a price of maintaining same interfaces and not incorporate all the edge cases that folks have already included. I will have to think about implementation but I agree that feels more _correct_ in the sense of probability

> Testing old versions of direct dependencies with modern versions of their recursive dependencies is not a useful or realistic check, in my opinion

Yes and no. 
- Yes because it's very unlikely to have direct dependency on pkg of >10 years old alongside indirect dependency of few months old. 
- No because it's technically possible hence some safeguards are required. 
- No because it's quite likely if we are thinking about relatively new packages with many releases, e.g. in our field: citril imports chevron of version (newest - 3 releases) and we want to use the newest rtables. I can definitely imagine this scenario.

I am going to create a new slide to discuss it.